dev/tools/pre-commit

```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+
+# Copy to .git/hooks/ to use.
+
+set -e
+
+CODE=0
+git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh --fix || CODE=1
+
+if [ $CODE -ne 0 ]
+then
+    2>&1 echo "Some files had newline errors; they have been fixed in the working tree."
+    2>&1 echo "Make sure to add them before committing."
+    2>&1 echo "This may fix itself if you were using git commit -a, and you try again."
+    exit 1
+fi
+
+if git diff-index --check --cached HEAD >/dev/null 2>&1 ;
+then
+    :
+else
+    2>&1 echo "Auto fixing whitespace issues..."
+
+    TEMP=$(mktemp)
+    # We fix whitespace in the index and in the working tree
+    # separately to preserve non-added changes.
+    git diff-index -p --cached HEAD > "$TEMP"
+    git apply --cached -R "$TEMP"
+    git apply --cached --whitespace=fix "$TEMP"
+
+    git diff-index -p HEAD > "$TEMP"
+    git apply -R "$TEMP"
+    git apply --whitespace=fix "$TEMP"
+    rm "$TEMP"
+
+    # Check that we did fix whitespace
+    if git diff-index --check --cached HEAD;
+    then
+        :
```

Comment:
Does all POSIX shell have `!`? In that case, you could also just invert the logic of the conditional.

Reply:
>Does all POSIX shell have `!`? In that case, you could also just invert the logic of the conditional.

Done.